### PR TITLE
Menu bar aggregation

### PR DESCRIPTION
**Pull request by DanVanAtta**
Builds on top of #841 
* Breaks up the two menu bar classes into individual classes, grouped by menu.
* Cleaned up the code between menu items, notably:
 - fixed hardcoding of substance class names

new_packages

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/844)

### CODE REVIEW

**Comment by DanVanAtta** [comment]
Similar @ron-murhammer  - please review this after #841, lots of potential for merge conflicts here.

**Comment by RoiEXLab** [comment]
@DanVanAtta cannot `import com.sun.deploy.util.SessionState;` , because the package `com.sun.deploy.util` does not exist...

**Comment by DanVanAtta** [comment]
fixed

**Comment by ron-murhammer** [comment]
@DanVanAtta It appears this already has merge conflicts.

**Comment by RoiEXLab** [comment]
@ron-murhammer this is due to the merge of #841 simple rebasing should be enough

**Comment by DanVanAtta** [comment]
now rebased

**Comment by ron-murhammer** [comment]
@DanVanAtta So looks like this PR is more than just moving methods to new classes. Its kind of hard to follow all the changes given the size of the diff. Thoughts on how best to review this? Are there any functional changes or just moving methods and refactoring?

File: src/games/strategy/triplea/ui/BasicGameMenuBar.java [hunk 1]
```diff
@@ -1,756 +0,0 @@
-package games.strategy.triplea.ui;
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/BasicGameMenuBar.java]: This class was chopped up. Look for the "extract" commits to see stuff being moved out.

**Comment by DanVanAtta** [comment]
No functionality changes. Previously the menu bar existed as two classes, one was a parent of the other.  The overall strategy was to re-slice these two classes, sending each menu to its own class. The commits are structured into roughly three phases:
1. Some initial maneuvering and cleanup
2. Breaking up the menus, begins with: "Extract new class 'ViewMenu' "
 - Reviewing these commits you should get the best view of what moved vs changed
3. Final cleanup after breaking up the menus, and then some additional house cleaning after the files were moved.

So, this can be reviewed commit by commit for sure to get a good view of the changes. Github on the other hand *is* doing a good job of showing what was modified in the moved files. The stuff that was a straight move can be identified from the very large contiguous blocks of removed and added code.

File: src/games/strategy/ui/SwingComponents.java [hunk 2]
```diff
@@ -20,6 +20,22 @@
 
 public class SwingComponents {
 
+  public enum KeyboardCode {
+    G(KeyEvent.VK_G);
```
Comment (ron-murhammer) [inline src/games/strategy/ui/SwingComponents.java]: Is it worth having an enum here?

**Comment by ron-murhammer** [comment]
@DanVanAtta Alright. Did a functionality test first and noticed 1 issue. The "Lobby" menu appears in singleplayer games with no dropdown options when it shouldn't. I'll start reviewing the commits now.

File: src/games/strategy/triplea/ui/TripleAFrame.java [hunk 3]
```diff
@@ -557,8 +557,6 @@ public void stopGame() {
     if (uiContext == null) {
       return;
     }
-    menu.dispose();
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/TripleAFrame.java]: Is this method call not needed?

File: src/games/strategy/ui/SwingComponents.java [hunk 4]
```diff
@@ -21,6 +21,7 @@
 public class SwingComponents {
 
   public enum KeyboardCode {
+    D(KeyEvent.VK_D),
     G(KeyEvent.VK_G);
```
Comment (ron-murhammer) [inline src/games/strategy/ui/SwingComponents.java]: Is this enum worth it?

**Comment by ron-murhammer** [comment]
@DanVanAtta Looks pretty good. Just a few minor comments and the 1 functional issue.

**Comment by DanVanAtta** [comment]
Fixed merge conflicts, added in a new commit "6056e78", now looking at the lobby menu issue.

File: src/games/strategy/triplea/ui/TripleAFrame.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/TripleAFrame.java]: It's not a call to a swing method. I thought it was.

So, removing this call means that the game notes window will not be closed when you close the main game window. But, we also have other windows that can be opened and are not included, so the game notes window is unique in having this behavior.

If we would like to restore this behavior, IMO we should cut an issue, I think we would largely be rewriting this, and referencing this diff is a good enough of a starting point. Meanwhile, the behavior of window closing is consistent (they stay open, unit help notes is an example)

File: src/games/strategy/ui/SwingComponents.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/ui/SwingComponents.java]: Yes! It'll grow as necessary.

Key benefits:
- we avoid the magic number interface swing uses. This class is evil: https://docs.oracle.com/javase/7/docs/api/javax/swing/SwingConstants.html, a number of swing interfaces are take multiple integer args, where each is a magic number. You always wonder, is it a magic number from the SwingConstants class? A subclass, or from the class you are using proper? 

So the enum gives us a strongly typed value, and we can avoid APIs that take in multiple int values, where each int means something different.

Second, we get trace ability. Doing a call hierarchy search on these reveals all usages. We'll be able to more easily answer the question of what shares which hot key. This also puts us on a path to replacing all of the specific keys with a user setting (namely buy consolidating usages and references to the keyboard keys, can eventually be more easily swapped out later)

**Comment by DanVanAtta** [comment]
Pushed a fix for the refactor bug:  https://github.com/triplea-game/triplea/pull/844/commits/6056e78c7c722013e78383cfaeb729c23288e551

Ready for another look

**Comment by ron-murhammer** [comment]
@DanVanAtta Merging now. Only comment is I'd prefer to have any PRs that are doing lots of moving methods/classes around just focused on that with no other changes. This just makes it much easier to review. In this particular case, I think having this PR broken into 2 or more PRs would have made it a bit easier as there was some other clean up and refactoring.
